### PR TITLE
sys/net/dhcpv6: Fixes for MUD URL option

### PR DESCRIPTION
**Pull request by JKRhb**
### Contribution description

This PR fixes a couple of mistakes I made in PR #15508 which added the Manufacturer Usage Description option (MUD, [RFC 8520](https://tools.ietf.org/html/rfc8520#section-10)) to the DHCPv6 client.

So far, the option was only included in Solicit Messages. According to [RFC 8415](https://tools.ietf.org/html/rfc8415#section-18.2.1), however, this is only permitted if it is "specifically allowed in the definition of individual options". I therefore removed the call of `_compose_mud_url_opt` from the `_solicit_servers` function and added it to `_request_renew_rebind` instead, including the option now in Request, Renew, and Rebind Messages. I updated the tests in `tests/gnrc_dhcpv6_client_6lbr` accordingly.

I also noticed that the naming of the Kconfig option I added in #15508 was incorrect. Setting the option `DHCPV6_CLIENT_MUD_URL` should now work as intended.

Furthermore, I rearranged some of the code, hopefully making the module a bit more efficient and robust. For example, the assertion of the MUD URL now happens in the `dhcpv6_client_init` function causing the application to crash immediately if something is wrong with the URL (rather than when a Request Message is sent).

Lastly I have noticed that the instructions for using the `tests-as-root` was outdated both in the readme files of  `tests/gnrc_dhcpv6_client_6lbr` and `tests/gnrc_dhcpv6_client`. I edited the two files in two separate commits which could also be merged into one (especially if this information should be missing in more readmes).

### Testing procedure

The changes to the client can be tested manually by using the updated tests provided in `tests/gnrc_dhcpv6_client_6lbr`. 

The Kconfig option can be tested, for instance, by first enabling Kconfig in the Makefile of `tests/gnrc_dhcpv6_client_6lbr` and then changing the MUD URL by using `make menuconfig`. If the URL is not `https://example.org` anymore, performing the tests with `sudo make test-as-root` should fail.

### Issues/PRs references

See also #15508.

### CODE REVIEW

File: sys/net/application_layer/dhcpv6/client.c [hunk 1]
```diff
@@ -127,6 +131,10 @@ static void *_thread(void *args)
 void dhcpv6_client_init(event_queue_t *eq, uint16_t netif)
 {
     assert(eq->waiter != NULL);
+    #ifdef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
```
Comment (benpicco) [inline sys/net/application_layer/dhcpv6/client.c:134]:
```suggestion
#ifdef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
```

File: sys/net/application_layer/dhcpv6/client.c [hunk 2]
```diff
@@ -127,6 +131,10 @@ static void *_thread(void *args)
...
+    assert(strlen(mud_url) <= MAX_MUD_URL_LENGTH);
+    assert(strncmp(mud_url, "https://", 8) == 0);
+    #endif  /* MODULE_GNRC_DHCPV6_CLIENT_MUD_URL */
```
Comment (benpicco) [inline sys/net/application_layer/dhcpv6/client.c:137]:
```suggestion
#endif  /* MODULE_GNRC_DHCPV6_CLIENT_MUD_URL */
```

File: sys/net/application_layer/dhcpv6/client.c [hunk 3]
```diff
@@ -247,8 +255,9 @@ static inline size_t _compose_elapsed_time_opt(dhcpv6_opt_elapsed_time_t *time)
     return len + sizeof(dhcpv6_opt_t);
 }
 
+#ifdef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
 static inline size_t _compose_mud_url_opt(dhcpv6_opt_mud_url_t *mud_url_opt,
-                                          const char *mud_url, size_t len_max)
+                                          size_t len_max)
 {
```
Comment (benpicco) [inline sys/net/application_layer/dhcpv6/client.c:261]: You can save an `#ifdef` down the line if this just returns 0 if the module is not used.

```suggestion
{
#ifndef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
    return 0;
#else
```

Comment (JKRhb) [reply sys/net/application_layer/dhcpv6/client.c:260]: Thanks for the suggestion! I added it down below :)

File: sys/net/application_layer/dhcpv6/client.c [hunk 4]
```diff
@@ -255,10 +255,12 @@ static inline size_t _compose_elapsed_time_opt(dhcpv6_opt_elapsed_time_t *time)
     return len + sizeof(dhcpv6_opt_t);
 }
 
-#ifdef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
 static inline size_t _compose_mud_url_opt(dhcpv6_opt_mud_url_t *mud_url_opt,
                                           size_t len_max)
 {
+#ifdef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
+    return 0;
```
Comment (benpicco) [inline sys/net/application_layer/dhcpv6/client.c:262]: Looks like you have that reversed :wink:

Comment (JKRhb) [reply sys/net/application_layer/dhcpv6/client.c]: Oops, classic mistake :D

File: sys/net/application_layer/dhcpv6/client.c [hunk 5]
```diff
@@ -248,8 +256,11 @@ static inline size_t _compose_elapsed_time_opt(dhcpv6_opt_elapsed_time_t *time)
 }
 
 static inline size_t _compose_mud_url_opt(dhcpv6_opt_mud_url_t *mud_url_opt,
-                                          const char *mud_url, size_t len_max)
+                                          size_t len_max)
 {
+#ifndef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
+    return 0;
+#endif /* MODULE_GNRC_DHCPV6_CLIENT_MUD_URL */
```
Comment (benpicco) [inline sys/net/application_layer/dhcpv6/client.c:263]: If that works you can also do

```suggestion
    if (!IS_USED(MODULE_GNRC_DHCPV6_CLIENT_MUD_URL)) {
        return 0;
    }
```

You'll have to remove the conditional definition of `mud_url`, but the optimizer should throw it out if it's not used.
You can check that the RAM/ROM size does not change.

Comment (JKRhb) [reply sys/net/application_layer/dhcpv6/client.c]: Looks good, I added it down below :)

**Comment by benpicco** [comment]
Please squash, let's see what Murdock thinks of this.

File: sys/net/application_layer/dhcpv6/client.c [hunk 6]
```diff
@@ -127,6 +129,10 @@ static void *_thread(void *args)
 void dhcpv6_client_init(event_queue_t *eq, uint16_t netif)
 {
     assert(eq->waiter != NULL);
+#ifdef MODULE_GNRC_DHCPV6_CLIENT_MUD_URL
+    assert(strlen(mud_url) <= MAX_MUD_URL_LENGTH);
+    assert(strncmp(mud_url, "https://", 8) == 0);
+#endif  /* MODULE_GNRC_DHCPV6_CLIENT_MUD_URL */
```
Comment (miri64) [inline sys/net/application_layer/dhcpv6/client.c:135]: Shouldn't

```suggestion
    if (IS_USED(MODULE_GNRC_DHCPV6_CLIENT_MUD_URL)) {
        assert(strlen(mud_url) <= MAX_MUD_URL_LENGTH);
        assert(strncmp(mud_url, "https://", 8) == 0);
    }
```

work here as well?

File: sys/net/application_layer/dhcpv6/client.c [hunk 7]
```diff
@@ -841,6 +841,8 @@ static void _request_renew_rebind(uint8_t type)
     if (type != DHCPV6_REBIND) {
         msg_len += _compose_sid_opt((dhcpv6_opt_duid_t *)&send_buf[msg_len]);
     }
+    msg_len += _compose_mud_url_opt((dhcpv6_opt_mud_url_t *)&send_buf[msg_len],
+                                     sizeof(send_buf) - msg_len);
```
Comment (miri64) [inline sys/net/application_layer/dhcpv6/client.c:845]: I think the indentation is off here a bit

```suggestion
    msg_len += _compose_mud_url_opt((dhcpv6_opt_mud_url_t *)&send_buf[msg_len],
                                    sizeof(send_buf) - msg_len);
```

**Review by miri64 — COMMENTED** [review]
Some last minute style comments ;-)

**Comment by miri64** [comment]
You may squash those immediately.

File: tests/gnrc_dhcpv6_client_6lbr/README.md [hunk 8]
```diff
@@ -24,7 +24,7 @@ The tests succeeds if you see the string `SUCCESS`.
 
 If any problems are encountered (i.e. if the test prints the string `FAILED`),
 set the echo parameter in the `run()` function at the bottom of the test script
-(tests/01-run.py) to `True`. The test script will then offer a more detailed
+(tests-as-root/01-run.py) to `True`. The test script will then offer a more detailed
```
Comment (JKRhb) [inline tests/gnrc_dhcpv6_client_6lbr/README.md:27]: @miri64 Thanks :) While I was at it I corrected the path to the test files in this line and also squashed it already. I hope that this was okay

**Review by miri64 — CHANGES_REQUESTED** [review]
When running the test on `samr21-xpro`, I get a timeout. When checking out e0f457 (the merge base of this PR) I do not get the timeout :-/

**Comment by miri64** [comment]
<details>

```
[mlenders@sarajevo RIOT]<3 BOARD=samr21-xpro make -C tests/gnrc_dhcpv6_client_6lbr/ -j test-as-root
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/gnrc_dhcpv6_client_6lbr'
make[1]: Nothing to be done for 'all'.
[sudo] password for mlenders: 


make[2]: Nothing to be done for 'all'.
sudo /home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/ethos/ethos tap0 /dev/ttyACM0 115200 
----> ethos: sending hello.
----> ethos: activating serial pass through.
DHCPv6: Selecting interface 7 as upstream
main()----> ethos: hello reply received
> ifconfig
help

> 
> ifconfig
Iface  7  HWaddr: BE:D6:64:3B:38:E9 
          L2-PDU:1500  MTU:1500  HL:64  RTR  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::bcd6:64ff:fe3b:38e9  scope: link  VAL
          inet6 addr: fe80::2  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff3b:38e9
          inet6 group: ff02::1:ff00:2
          
Iface  6  HWaddr: 3C:97  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
          
          Long HWaddr: 00:04:25:19:18:01:BC:97 
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::204:2519:1801:bc97  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff01:bc97
          
> help
Command              Description
---------------------------------------
reboot               Reboot the node
7
ifconfig 7
version              Prints current RIOT_VERSION
pm                   interact with layered PM subsystem
ps                   Prints information about running threads.
random_init          initializes the PRNG
random_get           returns 32 bit of pseudo randomness
nib                  Configure neighbor information base
ifconfig             Configure network interfaces
6ctx                 6LoWPAN context configuration tool
> ifconfig 7
Iface  7  HWaddr: BE:D6:64:3B:38:E9 
ifconfig
help
          L2-PDU:1500  MTU:1500  HL:64  RTR  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::bcd6:64ff:fe3b:38e9  scope: link  VAL
          inet6 addr: fe80::2  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff3b:38e9
          inet6 group: ff02::1:ff00:2
          
> ifconfig
Iface  7  HWaddr: BE:D6:64:3B:38:E9 
          L2-PDU:1500  MTU:1500  HL:64  RTR  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::bcd6:64ff:fe3b:38e9  scope: link  VAL
          inet6 addr: fe80::2  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff3b:38e9
          inet6 group: ff02::1:ff00:2
          
Iface  6  HWaddr: 3C:97  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
          
          Long HWaddr: 00:04:25:19:18:01:BC:97 
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::204:2519:1801:bc97  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff01:bc97
          
> help
Command              Description
---------------------------------------
reboot               Reboot the node
ifconfig 6
version              Prints current RIOT_VERSION
pm                   interact with layered PM subsystem
ps                   Prints information about running threads.
random_init          initializes the PRNG
random_get           returns 32 bit of pseudo randomness
nib                  Configure neighbor information base
ifconfig             Configure network interfaces
6ctx                 6LoWPAN context configuration tool
> ifconfig 6
Iface  6  HWaddr: 3C:97  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
          
          Long HWaddr: 00:04:25:19:18:01:BC:97 
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::204:2519:1801:bc97  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff01:bc97
          
> Timeout in expect script at "child.expect(r"inet6 addr:\s+{}[0-9a-fA-F:]+\s"" (tests/gnrc_dhcpv6_client_6lbr/tests-as-root/01-run.py:179)

make: *** [/home/mlenders/Repositories/RIOT-OS/RIOT/makefiles/tests/tests.inc.mk:35: test-as-root] Error 1
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/gnrc_dhcpv6_client_6lbr'
```
</details>

**Comment by JKRhb** [comment]
> When running the test on `samr21-xpro`, I get a timeout. When checking out e0f457 (the merge base of this PR) I do not get the timeout :-/

Hmm, that is strange :/ I just ran the test on native again and it succeeded – do you see indications on why it might fail on `samr21-xpro`?

**Comment by miri64** [comment]
> Hmm, that is strange :/ I just ran the test on native again and it succeeded – do you see indications on why it might fail on `samr21-xpro`?

The address seems not to be configured correctly. Maybe the buffer is too small?

**Comment by JKRhb** [comment]
> > Hmm, that is strange :/ I just ran the test on native again and it succeeded – do you see indications on why it might fail on `samr21-xpro`?
> 
> The address seems not to be configured correctly. Maybe the buffer is too small?

Hmm, do you mean the receive buffer? Or is there another buffer that could be affected?

I just ran the tests on an ESP32 where they also passed – could you maybe try it without once more without the MUD option for a comparison? (Sorry that this small change causes so much trouble :/)

**Comment by miri64** [comment]
> Hmm, do you mean the receive buffer? Or is there another buffer that could be affected?

Yeah, but sadly, increasing the receive buffer did not help :-/.

**Comment by miri64** [comment]
> I just ran the tests on an ESP32 where they also passed – could you maybe try it without once more without the MUD option for a comparison? (Sorry that this small change causes so much trouble :/)

With everything regarding MUD options commented out, the test succeeds :-/.

**Comment by JKRhb** [comment]
> 
> 
> When running the test on `samr21-xpro`, I get a timeout. When checking out e0f457 (the merge base of this PR) I do not get the timeout :-/

Hmm, that is strange :/ Do the 

> 
> 
> > I just ran the tests on an ESP32 where they also passed – could you maybe try it without once more without the MUD option for a comparison? (Sorry that this small change causes so much trouble :/)
> 
> With everything regarding MUD options commented out, the test succeeds :-/.

Hmm, very strange :/ Could you maybe run the rest with `ENABLE_DEBUG 1` to see where it gets stuck?

**Comment by miri64** [comment]
> > With everything regarding MUD options commented out, the test succeeds :-/.
> 
> Hmm, very strange :/ Could you maybe run the rest with `ENABLE_DEBUG 1` to see where it gets stuck?

Did this already, but doesn't really give helpful output. :(

**Comment by miri64** [comment]
Ah, gave it another try:

```
DHCPv6 client: received REPLY
DHCPv6 client: message is not from my server
```

**Comment by miri64** [comment]
As far as I can tell, there is a buffer overflow happening somewhere that overwrites the first few bytes of `server.duid`.

**Comment by miri64** [comment]
Ahhh, I believe `strncpy()` is provided the wrong length!

```diff
diff --git a/sys/net/application_layer/dhcpv6/client.c b/sys/net/application_layer/dhcpv6/client.c
index ce20966dcb..052935847f 100644
--- a/sys/net/application_layer/dhcpv6/client.c
+++ b/sys/net/application_layer/dhcpv6/client.c
@@ -262,13 +262,13 @@ static inline size_t _compose_mud_url_opt(dhcpv6_opt_mud_url_t *mud_url_opt,
     uint16_t len = strlen(mud_url);
 
     if (len > len_max) {
-        assert(0);
+        assert(len <= len_max);
         return 0;
     }
 
     mud_url_opt->type = byteorder_htons(DHCPV6_OPT_MUD_URL);
     mud_url_opt->len = byteorder_htons(len);
-    strncpy(mud_url_opt->mudString, mud_url, len_max);
+    strncpy(mud_url_opt->mudString, mud_url, len);
     return len + sizeof(dhcpv6_opt_mud_url_t);
 }
 
```

fixes it for me! (note that the `assert` is optional, but I think it makes the problem clearer in case the condition of the `assert()` is printed).

**Comment by miri64** [comment]
This bug existed previously, but probably adding it to the other message types revealed the bug more clearly.

**Comment by miri64** [comment]
also just noticed, that `mudString` is not in accordance with our coding conventions. We use `snake_case` not `camelCase`.

**Comment by JKRhb** [comment]
Thank you for identifying these issues! :) I pushed a new fixup commit with your suggestions

> also just noticed, that `mudString` is not in accordance with our coding conventions. We use `snake_case` not `camelCase`.

You are right, we used the casing from the RFC which was not the right choice. This should now also be fixed :)

**Review by miri64 — APPROVED** [review]
ACK, tested in on `samr21-xpro` and `native`. Both succeed now. To be on the save side I also ran the test with `make all-asan` on `native` which yielded nothing.

**Comment by miri64** [comment]
Please squash! :-)

**Comment by JKRhb** [comment]
Done :) Thank you once more for your patience and expertise :)

**Comment by miri64** [comment]
> You are right, back then we used the casing from the RFC which was not the right choice. This should now also be fixed :)

But even there it is [`MUDstring`](https://tools.ietf.org/html/rfc8520#section-10), so yet another variant ;-).

**Comment by JKRhb** [comment]
> 
> 
> > You are right, back then we used the casing from the RFC which was not the right choice. This should now also be fixed :)
> 
> But even there it is [`MUDstring`](https://tools.ietf.org/html/rfc8520#section-10), so yet another variant ;-).

Oh, ahem, then I made two mistakes at once – I'll try my best to bring it down to zero when I open the next PR ;)

**Comment by JKRhb** [comment]
Hmm, some tests on Murdock failed (see, e. g., [https://ci.riot-os.org/RIOT-OS/RIOT/15939/c2e1c0d3fca18f3225af12fca960473b7c0b2a1b/output/compile/tests/gnrc_dhcpv6_client_6lbr/msb-430h:gnu.txt](this one)). I guess this can be easily fixed by replacing `strncpy` with `memcpy`?

**Comment by miri64** [comment]
> Hmm, some tests on Murdock failed (see, e. g., [this one](https://ci.riot-os.org/RIOT-OS/RIOT/15939/c2e1c0d3fca18f3225af12fca960473b7c0b2a1b/output/compile/tests/gnrc_dhcpv6_client_6lbr/msb-430h:gnu.txt)). I guess this can be easily fixed by replacing `strncpy` with `memcpy`?

probably, yes. Though I'm wondering what GCC is actually trying to say with that error message ^^"

**Comment by miri64** [comment]
You can squash that already :-)

**Review by maribu — CHANGES_REQUESTED** [review]
Let's wait for https://github.com/RIOT-OS/RIOT/pull/15974 to get in and another round of Murdock afterwards before merging.

**Comment by miri64** [comment]
https://github.com/RIOT-OS/RIOT/pull/15974 was merged, so can you lift your block, please, @maribu?

**Pull request merged**